ext/datasource/refreshable_file/file_datasource.go

```diff
@@ -0,0 +1,108 @@
+package refreshable_file
+
+import (
+	"github.com/alibaba/sentinel-golang/ext/datasource"
+	"github.com/alibaba/sentinel-golang/logging"
+	"github.com/alibaba/sentinel-golang/util"
+	"github.com/fsnotify/fsnotify"
+	"github.com/pkg/errors"
+	"io/ioutil"
+	"os"
+	"time"
+)
+
+var (
+	logger = logging.GetDefaultLogger()
+)
+
+type RefreshableFileDataSource struct {
+	datasource.Base
+	sourceFilePath string
+	isInitialized  util.AtomicBool
+	watcher        *fsnotify.Watcher
+}
+
+func FileDataSourceStarter(sourceFilePath string, handlers ...datasource.PropertyHandler) (*RefreshableFileDataSource, error) {
+	var ds = &RefreshableFileDataSource{
+		sourceFilePath: sourceFilePath,
+	}
+	for _, h := range handlers {
+		ds.AddPropertyHandler(h)
+	}
+	return ds, ds.Initialize()
+}
+
+func (s *RefreshableFileDataSource) ReadSource() ([]byte, error) {
+	f, err := os.Open(s.sourceFilePath)
+	defer f.Close()
```

Comment:
This should be put after the error checking?

Reply:
Right